packages/astro/src/content/runtime.ts

```diff
@@ -319,9 +334,91 @@ type RenderResult = {
 	remarkPluginFrontmatter: Record<string, any>;
 };
 
+const IMAGE_REGEX = /__ASTRO_IMAGE_="([^"]+)"/g;
+
+async function updateImageReferencesInBody(html: string, fileName: string) {
+	// @ts-expect-error Virtual module
+	const { default: imageAssetMap } = await import('astro:asset-imports');
+
+	const imageObjects = new Map<string, GetImageResult>();
+
+	// @ts-expect-error Virtual module resolved at runtime
+	const { getImage } = await import('astro:assets');
+
+	// First load all the images. This is done outside of the replaceAll
+	// function because getImage is async.
+	for (const match of html.matchAll(IMAGE_REGEX)) {
+		const imagePath = match[1];
+		const decodedImagePath = JSON.parse(imagePath.replace(/&#x22;/g, '"'));
+		const id = imageSrcToImportId(decodedImagePath.src, fileName);
+
+		if (!id || imageObjects.has(id)) {
+			continue;
+		}
+
+		const imported = imageAssetMap.get(id);
+		if (!imported) {
+			continue;
+		}
+		const image: GetImageResult = await getImage({ src: imported });
+		imageObjects.set(id, image);
+	}
+
+	return html.replaceAll(IMAGE_REGEX, (full, imagePath) => {
+		const decodedImagePath = JSON.parse(imagePath.replace(/&#x22;/g, '"'));
+		const id = imageSrcToImportId(decodedImagePath.src, fileName);
+
+		if (!id) {
+			return full;
+		}
+
+		const image = imageObjects.get(id);
+
+		if (!image) {
+			return full;
+		}
+
+		return Object.entries({
+			...image.attributes,
+			src: image.src,
+			srcset: image.srcSet.attribute,
+			alt: decodedImagePath.alt,
```

Comment:
Won't the `alt` be part of `image.attributes`?

Reply:
No, because getImage is only passed the ESM import. Is that likely to be a problem? I wanted to be able to share one import between all of the instances of the image on the page, even if they had different alts.